ext/mbstring/libmbfl/filters/mbfilter_sjis_2004.c

```diff
@@ -659,3 +698,231 @@ int mbfl_filt_conv_wchar_jis2004_flush(mbfl_convert_filter *filter)
 
 	return 0;
 }
+
+static size_t mb_eucjp2004_to_wchar(unsigned char **in, size_t *in_len, uint32_t *buf, size_t bufsize, unsigned int *state)
+{
+	unsigned char *p = *in, *e = p + *in_len;
+	uint32_t *out = buf, *limit = buf + bufsize - 1;
+
+	while (p < e && out < limit) {
+		unsigned char c = *p++;
+
+		if (c <= 0x7F) {
+			*out++ = c;
+		} else if (c >= 0xA1 && c <= 0xFE) {
+			/* Kanji */
+			if (p == e) {
+				*out++ = MBFL_BAD_INPUT;
+				break;
+			}
+			unsigned char c2 = *p++;
+			if (c2 <= 0xA0 || c2 == 0xFF) {
+				*out++ = MBFL_BAD_INPUT;
+				continue;
+			}
+
+			unsigned int s1 = c - 0x80, s2 = c2 - 0x80;
+			unsigned int w1 = (s1 << 8) | s2, w = 0;
+
+			/* Conversion for combining characters */
+			if ((w1 >= 0x2477 && w1 <= 0x2479) || (w1 >= 0x2479 && w1 <= 0x247B) || (w1 >= 0x2577 && w1 <= 0x257E) || w1 == 0x2678 || w1 == 0x2B44 || (w1 >= 0x2B48 && w1 <= 0x2B4F) || (w1 >= 0x2B65 && w1 <= 0x2B66)) {
+				int k = mbfl_bisec_srch2(w1, jisx0213_u2_key, jisx0213_u2_tbl_len);
+				if (k >= 0) {
+					*out++ = jisx0213_u2_tbl[2*k];
+					*out++ = jisx0213_u2_tbl[2*k+1];
+					continue;
+				}
+			}
+
+			/* Conversion for BMP  */
+			w1 = (s1 - 0x21)*94 + s2 - 0x21;
+			if (w1 < jisx0213_ucs_table_size) {
+				w = jisx0213_ucs_table[w1];
+			}
+
+			/* Conversion for CJK Unified Ideographs ext.B (U+2XXXX) */
+			if (!w) {
+				w1 = (s1 << 8) | s2;
+				int k = mbfl_bisec_srch2(w1, jisx0213_jis_u5_key, jisx0213_u5_tbl_len);
+				if (k >= 0) {
+					w = jisx0213_jis_u5_tbl[k] + 0x20000;
+				}
+			}
+
+			*out++ = w ? w : MBFL_BAD_INPUT;
+		} else if (c == 0x8E && p < e) {
+			/* Kana */
+			unsigned char c2 = *p++;
+			if (c2 >= 0xA1 && c2 <= 0xDF) {
+				*out++ = 0xFEC0 + c2;
+			} else {
+				*out++ = MBFL_BAD_INPUT;
+			}
+		} else if (c == 0x8F && p < e) {
+			unsigned char c2 = *p++;
+			if ((c2 == 0xA1 || (c2 >= 0xA3 && c2 <= 0xA5) || c2 == 0xA8 || (c2 >= 0xAC && c2 <= 0xAF) || (c2 >= 0xEE && c2 <= 0xFE)) && p < e) {
+				unsigned char c3 = *p++;
+
+				if (c3 < 0xA1 || c3 == 0xFF) {
+					*out++ = MBFL_BAD_INPUT;
+					break;
```

Comment:
Should be continue?

Reply:
I think you're right! Good catch!

I don't think this will cause any observable difference in behavior; just a tiny loss of performance between the conversion function will return a half-full buffer to the caller, and then the caller will have to re-enter it again.

Reply:
Fixed.